modules/oe_editorial_unpublish/tests/Functional/EditorialUnpublishTest.php

```diff
@@ -0,0 +1,170 @@
+<?php
+
+declare(strict_types = 1);
+
+namespace Drupal\Tests\oe_editorial_unpublish\Functional;
+
+use Drupal\Core\Url;
+use Drupal\Tests\BrowserTestBase;
+
+/**
+ * Tests the unpublishing form for nodes.
+ */
+class EditorialUnpublishTest extends BrowserTestBase {
+
+  /**
+   * The entity type manager.
+   *
+   * @var \Drupal\node\NodeStorageInterface
+   */
+  protected $nodeStorage;
+
+  /**
+   * The test node.
+   *
+   * @var \Drupal\node\NodeInterface
+   */
+  protected $node;
+
+  /**
+   * The current user.
+   *
+   * @var \Drupal\user\UserInterface
+   */
+  protected $user;
+
+  /**
+   * {@inheritdoc}
+   */
+  protected static $modules = [
+    'user',
+    'node',
+    'system',
+    'oe_editorial',
+    'oe_editorial_corporate_workflow',
+    'oe_editorial_workflow_demo',
+    'oe_editorial_unpublish_test',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setUp() {
+    parent::setUp();
+
+    $entity_type_manager = $this->container->get('entity_type.manager');
+    $this->nodeStorage = $entity_type_manager->getStorage('node');
+    $this->node = $this->nodeStorage->create(
+      [
+        'type' => 'oe_workflow_demo',
+        'title' => 'My node',
+        'moderation_state' => 'draft',
+      ]
+    );
+    $this->node->save();
+
+    /** @var \Drupal\user\RoleInterface $role */
+    $role = $entity_type_manager->getStorage('user_role')->load('oe_validator');
+    $this->user = $this->drupalCreateUser($role->getPermissions());
+    $this->drupalLogin($this->user);
+  }
+
+  /**
+   * Tests access to the unpublishing form.
+   */
+  public function testUnpublishAccess(): void {
+    $unpublish_url = Url::fromRoute('entity.node.unpublish', [
+      'node' => $this->node->id(),
+    ]);
+
+    // Assert that we can't access the unpublish page when the node is not
+    // published.
+    $this->assertFalse($unpublish_url->access($this->user));
+
+    // Publish the page.
+    $this->node->moderation_state->value = 'published';
+    $this->node->save();
+
+    // A user with permissions can access the unpublish page.
+    $this->assertTrue($unpublish_url->access($this->user));
+
+    // A user without permissions can not access the unpublish page.
+    $this->drupalLogout();
+    $this->drupalGet($unpublish_url);
+    $this->assertSession()->statusCodeEquals(403);
+
+    // We can't access the unpublish page if the last revision is not published.
+    $this->drupalLogin($this->user);
+    $this->node->moderation_state->value = 'draft';
+    $this->node->save();
+    $this->assertFalse($unpublish_url->access($this->user));
```

Comment:
Currently this is true because of the nature of the way corporate workflow is defined by business. However it should not be a representative hard rule as I said above. I would remove this part.

Reply:
In the test it doesn't matter. The test is using the corporate workflow so we need to assert that the access depends on that. Having removed that in the access check still would not be possible due to lack of transitions. Also, we kept the access check of `!$this->moderationInfo->isDefaultRevisionPublished($entity)`